src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitService.java

```diff
@@ -82,28 +110,62 @@ public UUID runAndSaveResult(UUID networkUuid, String variantId, String receiver
         List<String> variableTwoWindingsTransformers = new ArrayList<>();
         List<String> variableShuntCompensators = new ArrayList<>();
+
         voltageInitParametersEntity.ifPresent(voltageInitParameters -> {
             if (voltageInitParameters.getVoltageLimits() != null) {
-                voltageInitParameters.getVoltageLimits().forEach(voltageLimit -> {
-                    var filterEquipments = filterService.exportFilters(voltageLimit.getFilters().stream().map(FilterEquipmentsEmbeddable::getFilterId).toList(), networkUuid, variantId);
-                    filterEquipments.forEach(filterEquipment ->
-                            filterEquipment.getIdentifiableAttributes().forEach(idenfiableAttribute -> {
-                                    specificVoltageLimits.add(new VoltageLimitOverride(idenfiableAttribute.getId(), VoltageLimitType.LOW_VOLTAGE_LIMIT, true, voltageLimit.getLowVoltageLimit()));
-                                    specificVoltageLimits.add(new VoltageLimitOverride(idenfiableAttribute.getId(), VoltageLimitType.HIGH_VOLTAGE_LIMIT, true, voltageLimit.getHighVoltageLimit()));
-                                }
-                            )
-                    );
-                });
-                constantQGenerators.addAll(toEquipmentIdsList(voltageInitParameters.getConstantQGenerators(), networkUuid, variantId));
-                variableTwoWindingsTransformers.addAll(toEquipmentIdsList(voltageInitParameters.getVariableTwoWindingsTransformers(), networkUuid, variantId));
-                variableShuntCompensators.addAll(toEquipmentIdsList(voltageInitParameters.getVariableShuntCompensators(), networkUuid, variantId));
+                Network network = getNetwork(networkUuid, PreloadingStrategy.COLLECTION, variantId);
+
+                Map<String, VoltageLimitEntity> voltageLevelDefaultLimits = resolveVoltageLevelLimits(voltageInitParameters.getVoltageLimits().stream().filter(voltageLimit -> VoltageLimitParameterType.DEFAULT.equals(voltageLimit.getVoltageLimitParameterType())).toList(), networkUuid, network.getVariantManager().getWorkingVariantId());
+                Map<String, VoltageLimitEntity> voltageLevelModificationLimits = resolveVoltageLevelLimits(voltageInitParameters.getVoltageLimits().stream().filter(voltageLimit -> VoltageLimitParameterType.MODIFICATION.equals(voltageLimit.getVoltageLimitParameterType())).toList(), networkUuid, network.getVariantManager().getWorkingVariantId());
+                List<VoltageLevel> voltageLevels = network.getVoltageLevelStream()
+                    .filter(voltageLevel -> voltageLevelDefaultLimits.keySet().contains(voltageLevel.getId()) || voltageLevelModificationLimits.keySet().contains(voltageLevel.getId()))
+                    .toList();
+                voltageLevels.forEach(voltageLevel -> fillSpecificVoltageLimits(specificVoltageLimits, voltageLevelModificationLimits, voltageLevelDefaultLimits, voltageLevel));
```

Comment:
```suggestion
                network.getVoltageLevelStream()
                    .forEach(voltageLevel -> {
                        if (voltageLevelDefaultLimits.keySet().contains(voltageLevel.getId()) || voltageLevelModificationLimits.keySet().contains(voltageLevel.getId()) {
                            fillSpecificVoltageLimits(specificVoltageLimits, voltageLevelModificationLimits, voltageLevelDefaultLimits, voltageLevel);
                        }
                    });
```
instead of multiple loop no ?

Reply:
Agreed, i just kept the filter instead of the if statement